### ISSUE TITLE
Can't use function overloads

### ISSUE THREAD

**Issue by ttttttssssss** [typescript-eslint/typescript-eslint#3595]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
// .eslintrc.js
module.exports = {
  root: true,
  env: {
    browser: true,
    es2021: true
  },
  parser: '@typescript-eslint/parser',
  extends: [
    'standard'
  ],
  parserOptions: {
    ecmaVersion: 12,
    sourceType: 'module'
  },
  rules: {
    indent: [2, 2]
  }
}

```

```TS
// index.ts
function log(x: string): string
function log(x: number): number
function log (x: number | string): number | string {
  if (typeof x === 'number') {
    return x
  } else {
    return x
  };
}

log(1)

```

**Expected Result**

**Actual Result**
![image](https://user-images.githubusercontent.com/58020640/123927671-a9761080-d9bf-11eb-970c-07657fab58d1.png)

```
// throw error
'log' is already defined.
'log' is already defined.
```

**Additional Info**
dir
```
.
├── index.ts
├── node_modules
├── .eslintrc.js
├── package-lock.json
└── package.json
```
package.json
```json
{
  "name": "ts",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "typescript": "^4.3.4",
    "@typescript-eslint/parser": "^4.28.1",
    "eslint": "^7.29.0",
    "eslint-config-standard": "^16.0.3",
    "eslint-plugin-import": "^2.23.4",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-promise": "^4.3.1",
    "eslint-webpack-plugin": "^2.5.4"
  }
}

```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.28.1` |
| `TypeScript`                | `4.3.4` |
| `ESLint`                    | `7.29.0` |
| `node`                      | `12.19.0` |
| `vscode extension ESLint`                      | `2.1.23` |

**Comment by bradzacher** [comment]
> - [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

You ticked the box... But I don't think you actually read it.
The very first FAQ answers your question.
